### PR TITLE
Add UTF-16/UTF-8 byte offset conversion utilities

### PR DESCRIPTION
**Pull request by tilladam**
## Summary

- Adds `unicode_utils` module to `i-slint-core` consolidating duplicate UTF-16/UTF-8 offset conversion code
- Replaces inline implementations in `android-activity/javahelper.rs`, `qt/qt_window.rs`, and `core/items/text.rs` with calls to the shared module
- Byte offset boundary helpers (`ceil_byte_offset`, `floor_byte_offset`) replace a hand-rolled `floor_char_boundary` equivalent (stdlib version requires Rust 1.91, MSRV is 1.88)

## Functions

**Byte offset utilities:**
- `is_valid_byte_offset` — check if offset is on a UTF-8 character boundary
- `floor_byte_offset` / `ceil_byte_offset` — snap to nearest valid boundary
- `byte_offset_to_char_count` / `char_count_to_byte_offset` — character count conversions

**UTF-16 conversions (needed for Android InputConnection + iOS UITextInput):**
- `utf16_offset_to_byte_offset` — UTF-16 → UTF-8 (`None` for invalid mid-surrogate offsets)
- `byte_offset_to_utf16_offset` — UTF-8 → UTF-16
- `utf16_offset_to_byte_offset_clamped` — UTF-16 → UTF-8 with forward clamping (matches original `convert_utf16_index_to_utf8` semantics)

## Test plan

- [x] 39 unit tests + 3 doctests covering ASCII, BMP, supplementary plane (emoji), combining characters, surrogate pair edge cases, empty strings, roundtrip conversions
- [x] `cargo build` for i-slint-core, i-slint-backend-qt, i-slint-backend-android-activity
- [x] `cargo test -p i-slint-core unicode_utils`

Split out from #10557 per review feedback to use smaller increments.

### CODE REVIEW

File: internal/core/items/text.rs [hunk 1]
```diff
@@ -1348,20 +1348,8 @@ fn safe_byte_offset(unsafe_byte_offset: i32, text: &str) -> usize {
...
-    if text.is_char_boundary(byte_offset_candidate) {
-        return byte_offset_candidate;
-    }
-
-    // Use std::floor_char_boundary once stabilized.
-    text.char_indices()
-        .find_map(|(offset, _)| if offset >= byte_offset_candidate { Some(offset) } else { None })
-        .unwrap_or(text.len())
+    // Use std::ceil_char_boundary once MSRV >= 1.91.
+    crate::unicode_utils::ceil_byte_offset(text, unsafe_byte_offset as usize)
```
Comment (tronical) [inline internal/core/items/text.rs]: There's only this one single call site here. I suggest to either keep the code here **and** the comment to replace this with `std::ceil_char_boundary` when we bump the MSRV, or have that comment in the `ceil_byte_offset` function.

File: internal/core/unicode_utils.rs [hunk 2]
```diff
@@ -0,0 +1,553 @@
...
+///
+/// // CJK: "日" is 3 UTF-8 bytes, 1 UTF-16 code unit
+/// assert_eq!(byte_offset_to_utf16_offset("日本", 3), 1);
+///
+/// // Emoji: "😀" is 4 UTF-8 bytes, 2 UTF-16 code units
+/// assert_eq!(byte_offset_to_utf16_offset("a😀b", 5), 3);
+/// ```
+pub fn byte_offset_to_utf16_offset(text: &str, byte_offset: usize) -> usize {
+    assert!(
+        is_valid_byte_offset(text, byte_offset),
```
Comment (tronical) [inline internal/core/unicode_utils.rs]: This is - as far as I can tell - the only call site for a function where I feel the one-liner it is (`offset <= text.len() && text.is_char_boundary(offset)`) is more readable than `is_valid_byte_offset()`.

I'd prefer to inline here.

However: Why is a panic here better than what I suspect would be "rounding up" behaviour?

File: internal/core/unicode_utils.rs [hunk 3]
```diff
@@ -0,0 +1,553 @@
...
+///
+/// // CJK: "日" is 3 UTF-8 bytes, 1 UTF-16 code unit
+/// assert_eq!(utf16_offset_to_byte_offset("日本", 1), Some(3));
+///
+/// // Emoji: "😀" is 4 UTF-8 bytes, 2 UTF-16 code units (surrogate pair)
+/// assert_eq!(utf16_offset_to_byte_offset("a😀b", 1), Some(1));
+/// assert_eq!(utf16_offset_to_byte_offset("a😀b", 2), None); // inside surrogate pair
+/// assert_eq!(utf16_offset_to_byte_offset("a😀b", 3), Some(5));
+/// ```
+pub fn utf16_offset_to_byte_offset(text: &str, utf16_offset: usize) -> Option<usize> {
```
Comment (tronical) [inline internal/core/unicode_utils.rs]: There's no call site for this function yet. I'd prefer to introduce this function in a PR that uses the code.

**Review by tronical — COMMENTED** [review]
I'm generally in favour of de-duplicating code when there's value - in terms of improving readability or unifying complexity. I have the feeling that the LSP might have similar code duplicated.

That said, I prefer introducing code when it's used.

So here the android and qt backend could benefit from a byte_offset_to_utf16_offset shared function, because that's shared. I even agree :-) that the expression here becomes more readable, i.e. `byte_offset_to_utf16_offset` is better than `in_str[..utf8_index].encode_utf16().count()`. We could bike-shed if we should call them indices or offsets, but I'm fine either way :-).

Also, while qt and android backends use different versions of the inverse (clamped vs. non-clamped), they could be seen as "siblings" and I think that's also fine to share in this PR.

However, for the other functions I'd prefer them to be introduced with changes that use them, so it's more apparent what the value is.

File: internal/core/unicode_utils.rs (hunk 2 above)
Comment (tronical) [reply internal/core/unicode_utils.rs]: Code wise this seems "resolved" in the sense of inlined, but I'm curious about what you think about my panic question :)

**Review by tronical — CHANGES_REQUESTED** [review]
AFAICS there are two unresolved comments :)

File: internal/core/unicode_utils.rs (hunk 2 above)
Comment (tilladam) [reply internal/core/unicode_utils.rs]: The callers (Qt and Android backends) pass byte offsets from `TextInput::cursor_position`, which are always valid UTF-8 boundaries. A panic here catches programming errors (e.g. arithmetic on byte offsets that lands mid-character) immediately rather than silently producing wrong cursor positions that would be hard to debug.

That said, if you'd prefer a non-panicking API that rounds up via `ceil_byte_offset`, I'm happy to change it. Let me know.

File: internal/core/items/text.rs (hunk 1 above)
Comment (tilladam) [reply internal/core/items/text.rs]: Good point. `ceil_byte_offset` is now also called internally by `byte_offset_to_utf16_offset` as its release-build fallback (see below), so it has two callers and the extraction earns its place. The MSRV note stays on the function.

File: internal/core/unicode_utils.rs (hunk 2 above)
Comment (tilladam) [reply internal/core/unicode_utils.rs]: The `is_valid_byte_offset` helper is gone — the check is now inlined as the `debug_assert!` condition in `byte_offset_to_utf16_offset`. For the panic question: changed to `debug_assert!` combined with a `ceil_byte_offset` call as the unconditional first step, so debug builds still catch invalid offsets loudly while release builds silently round up to the next valid boundary rather than crashing.

File: internal/core/unicode_utils.rs (hunk 3 above)
Comment (tilladam) [reply internal/core/unicode_utils.rs]: Addressed: the non-clamped `utf16_offset_to_byte_offset` (the one with no callers at review time) has been removed. All three remaining functions now have callers: `ceil_byte_offset` from `text.rs` and from `byte_offset_to_utf16_offset` internally, `byte_offset_to_utf16_offset` from the Qt and Android backends, and `utf16_offset_to_byte_offset_clamped` from the Android backend.

File: internal/core/unicode_utils.rs (hunk 2 above)
Comment (tilladam) [reply internal/core/unicode_utils.rs]: Thanks for the nudge. Addressed by using `debug_assert!` instead of `assert!`, combined with an unconditional `ceil_byte_offset` call before the conversion: debug builds still fire immediately on an invalid offset, and release builds silently round up to the next valid boundary.

**Comment by tilladam** [comment]
Following up on your observation about the LSP: `internal/compiler/diagnostics.rs` does have two inline `adjust_utf16` closures that implement the same algorithms as `byte_offset_to_utf16_offset` and `utf16_offset_to_byte_offset_clamped` — used in `SourceFile::line_column` and `SourceFile::offset` for LSP position conversion.

Consolidating that duplication would require `i-slint-compiler` to gain a dependency on `i-slint-core` (since the compiler currently only depends on `i-slint-common`), which seems like the wrong direction. A dedicated new crate for three functions isn't worth the overhead either, especially given that `ceil_byte_offset` will disappear when the MSRV reaches 1.91 and the UTF-16 functions are unlikely to grow much further.

The cleanest fix for the compiler side would be a small private helper in `diagnostics.rs` itself. Happy to do that as a follow-up if it's useful, but leaving it out of this PR to keep the scope focused.

File: internal/core/unicode_utils.rs
Comment (tronical) [inline internal/core/unicode_utils.rs:1]: After #11668, shouldn't this file be gone? :)

**Review by tronical — CHANGES_REQUESTED** [review]

File: internal/core/unicode_utils.rs
Comment (tilladam) [reply internal/core/unicode_utils.rs:1]: Good catch — addressed in bf7d02f76:

- Deleted `internal/core/unicode_utils.rs` (the duplicate)
- Removed `pub mod unicode_utils;` from `internal/core/lib.rs`
- Inlined `ceil_byte_offset` back into `text::safe_byte_offset` per your original Feb 27 suggestion (single call site, MSRV comment moves with it)
- Qt and Android backends now import from `i_slint_common::unicode_utils` directly
- Added `i-slint-common` to the Android backend's `Cargo.toml` (Qt already had it)

Verified with `cargo test -p i-slint-core` (248 pass), `cargo test -p i-slint-common` (5 pass — the unicode_utils tests there already cover ASCII/BMP/surrogate/roundtrip), and `cargo check` for the Qt and Android backends.

File: internal/core/items/text.rs [hunk 4]
```diff
@@ -1366,20 +1366,16 @@ fn safe_byte_offset(unsafe_byte_offset: i32, text: &str) -> usize {
...
-    if byte_offset_candidate >= text.len() {
+    // Use str::ceil_char_boundary once MSRV >= 1.91.
+    let offset = unsafe_byte_offset as usize;
+    if offset >= text.len() {
         return text.len();
     }
-
-    if text.is_char_boundary(byte_offset_candidate) {
-        return byte_offset_candidate;
+    let mut pos = offset;
```
Comment (tronical) [inline internal/core/items/text.rs]: I suppose the changes in this file can also be removed/undone?

That said, both functions, `ceil_char_boundary` and `floor_char_boundary` can be used now.

Comment (tilladam) [reply internal/core/items/text.rs]: Done in a23fb44da — `safe_byte_offset` now calls `text.ceil_char_boundary(...)` directly. MSRV is 1.92, so both `ceil_char_boundary` and `floor_char_boundary` are available; `ceil_char_boundary` matches the original round-up behaviour.

**Comment by tilladam** [comment]
@tronical follow-ups from the 2026-05-26 round are in:

- bf7d02f76: deleted the duplicate `internal/core/unicode_utils.rs` (made obsolete by #11668), removed its module declaration, inlined `ceil_byte_offset` back into `text::safe_byte_offset`.
- a23fb44da: switched `safe_byte_offset` to call `text.ceil_char_boundary(...)` directly (now stable since MSRV is 1.92).

PR is mergeable. Mind a re-review when you have a moment?

**Review by tronical — APPROVED** [review]
Thanks:)
